common/src/main/java/com/microsoft/identity/common/internal/broker/PackageHelper.java

```diff
@@ -0,0 +1,135 @@
+                    PackageManager.GET_SIGNATURES);
+            if (info != null && info.signatures != null && info.signatures.length > 0) {
+                Signature signature = info.signatures[0];
+                MessageDigest md = MessageDigest.getInstance("SHA");
+                md.update(signature.toByteArray());
+                return Base64.encodeToString(md.digest(), Base64.NO_WRAP);
+                // Server side needs to register all other tags. ADAL will
+                // send one of them.
+            }
+        } catch (NameNotFoundException e) {
+            Logger.error(TAG, "Calling App's package does not exist in PackageManager. ", "", e);
+        } catch (NoSuchAlgorithmException e) {
+            Logger.error(TAG, "Digest SHA algorithm does not exists. ", "", e);
+        }
+        return null;
+    }
+
+    /**
+     * Gets the kernel user-ID that has been assigned to this application.
+     *
+     * @param packageName for which the user id has to be returned
+     * @return UID user id
+     */
+    public int getUIDForPackage(final String packageName) {
+        int callingUID = 0;
+        try {
+            final ApplicationInfo info = mPackageManager.getApplicationInfo(packageName, 0);
+            if (info != null) {
+                callingUID = info.uid;
+            }
+        } catch (NameNotFoundException e) {
+            Logger.error(TAG, "Package is not found. ", "Package name: " + packageName, e);
+        }
+        return callingUID;
+    }
+
+    /**
+     * Gets redirect uri for broker.
+     *
+     * @param packageName     application package name
+     * @param signatureDigest application signature
+     * @return broker redirect url
+     */
+    public static String getBrokerRedirectUrl(final String packageName, final String signatureDigest) {
+        if (!StringExtensions.isNullOrBlank(packageName)
+                && !StringExtensions.isNullOrBlank(signatureDigest)) {
+            // If the caller is the Authenticator, then use the broker redirect URI.
+            if (packageName.equals(AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME) &&
+                    signatureDigest.equals(AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_SIGNATURE)) {
+                return AuthenticationConstants.Broker.BROKER_REDIRECT_URI;
+            }
+
+            try {
+                return String.format("%s://%s/%s", AuthenticationConstants.Broker.REDIRECT_PREFIX,
+                        URLEncoder.encode(packageName, AuthenticationConstants.ENCODING_UTF8),
+                        URLEncoder.encode(signatureDigest, AuthenticationConstants.ENCODING_UTF8));
+            } catch (UnsupportedEncodingException e) {
+                // This encoding issue will happen at the beginning of API call,
+                // if it is not supported on this device. ADAL uses one encoding
+                // type.
+                Logger.error(TAG, "", "Encoding is not supported", e);
```

Comment:
Second param to `Logger.error` can be omitted if no `correlation_id` is present. If there is a `correlation_id` set on the `DiagnosticContext` it will use that.